### PR TITLE
Add Moondream transformer implementation and example

### PR DESCRIPTION
**Pull request by santiagomed**
This PR adds a [moondream.ai](moondream.ai)'s Rust implementation using Candle to the `candle-transformers` crate as well as a usage example in `candle-examples` crate.

Changelog:
- Implement `moondream` in the `candle-transformers` crate. It includes the Moondream model consisting of a Vision Encoder and a Phi Mixformer model (from the `mixformer` transformer implementation).
- Add a `candle-moondream` example to the `candle-examples` crate. Consists of a CLI application to run the Moondream model based on CLI arguments
image
- Add a `forward_with_img` method to the `mixformer` implementation to allow the passing of image embeddings into the mode.

### CODE REVIEW

File: candle-transformers/src/models/moondream.rs [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+    patch_embed: LinearPatchEmbedding,
+    pos_embed: Tensor,
+    blocks: Vec<VitBlock>,
+    norm: candle_nn::LayerNorm,
+}
+
+impl VisionTransformer {
+    fn new(cfg: &VisionConfig, vb: VarBuilder) -> Result<Self> {
+        let embed_len = 729;
+        let embed_dim = 1152;
```
Comment (LaurentMazare) [inline candle-transformers/src/models/moondream.rs:153]: Feels that these hardcoded values should come from `VisionConfig`?

**Review by LaurentMazare — COMMENTED** [review]
Looks great, thanks a lot for making this.
Would you mind adding some instructions to the readme on how to run the model and the expected outputs? Also could you remove the images as the repo is already pretty large. For the examples you can either wget the images from some other source or use images that are already in the repo.

**Comment by santiagomed** [comment]
@LaurentMazare fixed a bug which was producing incorrect output and addressed your comments! Let me know if anything else is needed!

**Review by LaurentMazare — APPROVED** [review]

**Comment by LaurentMazare** [comment]
Perfect, thanks again!

**Pull request merged**